### DISCUSSION TITLE
Questions about torque input

### DISCUSSION THREAD

**Discussion by GichanCho** [raisimTech/raisimLib#200] [Q&A]
Hello, this is a student of Motion Control Lab, DGIST.

I try to apply PD closed-loop controller to 7-DoF manipulator. In that process, I have some questions.

![Block diagram](https://user-images.githubusercontent.com/90598192/136139076-a0286920-6953-409a-974f-fa48afcc2a0e.png)

1. In the attached figure, I want to know how to check 'tau-star(torque-input)' and 'theta-2dot(angular-acceleration)'. Is there any API or methods?

2. If I want to modify or directly input the 'tau-star' with real-time, I wonder how to do that.

3. The code attached below is a PD controller that currently used. And this is based on RaiSim manual.

// iiwa14_manipulator PD controller
    singleA->setGeneralizedCoordinate(initial_position);
    singleA->setGeneralizedForce(Eigen::VectorXd::Zero(singleA->getDOF()));
    singleA->setPdGains(jointPgain, jointDgain);
    singleA->setPdTarget(pTarget1, vTarget1);
    singleA->setName("iiwa14");

Is this closed-loop / feed back controller?

P.S.

안녕하세요 교수님, 위에 질문을 한 번 드렸지만, 제 부족한 영어실력으로 인해 정확한 의미가 전달되지 않을 수도 있어 이렇게 추신에 한글로 다시 질문을 남깁니다.

1. 첨부해드린 이미지에 보면, 일반적으로 position-tracking을 하는 경우 위치 reference가 controller를 거치면 plant에 들어가는 torque가 됩니다. 현재 이미지에는 tau*로 표기를 해두었는데, joint에 들어가는 torque를 제어하는 것이 main interest 입니다.

해서, tau*를 확인할 수 있는 방법이 있는지 궁금하고 또한 tau*가 joint 각각에 계산되어 들어가는지, output인 각가속도는 확인할 방법이 있는지 궁금합니다.

2. 또한 따로 controller를 이용하지 않고, user가 real-time으로 tau*를 넣어줄 수 있는 방법이 있는지 / 가능한지 궁금합니다.

3. RaiSim 홈페이지를 참고해서 현재 코드를 작성하였는데, 다른 조치 없이 저 상태 그대로 FB 제어기가 되는 것인지, 만약 DOB나 gravity compensation을 하고 싶다면 어떻게 응용해야하는지 질문드리고 싶습니다.

감사합니다.

**Comment by jhwangbo** [comment]
1. a) how to get the generalized force? ``getGeneralizedForce``()``
    b) how to get the generalized acceleration? There is no way to get the generalized acceleration. The best method to get the generalized acceleration is a finite difference.
2. ``setGeneralizedForce()``. You can call this function real-time. 
3.  the built-in pd controller is an implicit PD controller. It is a feedback controller. I don't quite get the question. You have to implement the DOB and gravity compensation and send the torque to the joints. The interface is trivial
